typescript-ignorant/TypeScriptParser.g4

```diff
@@ -0,0 +1,838 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ */
+parser grammar TypeScriptParser;
+
+options {
+    tokenVocab=TypeScriptLexer;
+    superClass=TypeScriptBaseParser;
+}
+
+// SupportSyntax
+
+initializer
+    : '=' singleExpression
+    ;
+
+bindingPattern
+    : (arrayLiteral | objectLiteral)
+    ;
+
+// TypeScript SPart
+// A.1 Types
+
+typeParameters
+    : '<' typeParameterList? '>'
+    ;
+
+typeParameterList
+    : typeParameter (',' typeParameter)*
+    ;
+
+typeParameter
+    : Identifier constraint?
+    | typeParameters
+    ;
+
+constraint
+    : Extends type_
+    ;
+
+typeArguments
+    : '<' typeArgumentList? '>'
+    ;
+
+typeArgumentList
+    : typeArgument (',' typeArgument)*
+    ;
+
+typeArgument
+    : type_
+    ;
+
+type_
+    : unionOrIntersectionOrPrimaryType
+    | conditionalType
+    | functionType
+    | constructorType
+    | typeGeneric
+    | StringLiteral
+    ;
+
+conditionalType
+    : primaryType Extends primaryType '?' type_ ':' type_
+    ;
+
+unionOrIntersectionOrPrimaryType
+    : unionOrIntersectionOrPrimaryType '|' unionOrIntersectionOrPrimaryType # Union
+    | unionOrIntersectionOrPrimaryType '&' unionOrIntersectionOrPrimaryType # Intersection
+    | primaryType # Primary
+    ;
+
+primaryType
+    : '(' type_ ')'                                 # ParenthesizedPrimType
+    | predefinedType                                # PredefinedPrimType
+    | typeReference                                 # ReferencePrimType
+    | objectType                                    # ObjectPrimType
+    | primaryType {notLineTerminator()}? '[' ']'    # ArrayPrimType
+    | '[' tupleElementTypes ']'                     # TuplePrimType
+    | typeQuery                                     # QueryPrimType
+    | This                                          # ThisPrimType
+    | typeReference Is primaryType                  # RedefinitionOfType
+    | Keyof primaryType                             # KeyofType
+    ;
+
+predefinedType
+    : Any
+    | Number
+    | Boolean
+    | String
+    | Symbol
+    | Void
+    | Never
+    ;
+
+typeReference
+    : typeName ( typeIncludeGeneric | typeGeneric | '[' Identifier ']')?
+    ;
+
+// I tried recursive include, but it's not working.
+// typeGeneric
+//    : '<' typeArgumentList typeGeneric?'>'
+//    ;
+//
+// TODO: Fix recursive
+//
+typeGeneric
+    : '<' typeArgumentList '>'
+    ;
+
+typeIncludeGeneric
+    :'<' typeArgumentList '<' typeArgumentList ('>' bindingPattern '>' | '>>')
```

Comment:
Maybe just remove `'>>'` token and use `'>'` everywhere for simplicity (as it works in [Java grammar](https://github.com/antlr/grammars-v4/tree/master/java/java))?

Reply:
I just copied it from the original `typescript` grammar. If we're to change this we'll have to test and fix it. Maybe later?

Reply:
It's related to the `typeGeneric` recursive problem.